library/src/scripts/components/siteNav/SiteNavNode.tsx

```diff
@@ -0,0 +1,240 @@
+/*
+ * @author Stéphane LaFlèche <stephane.l@vanillaforums.com>
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license GPL-2.0-only
+ */
+
+import * as React from "react";
+import classNames from "classnames";
+import { Link } from "react-router-dom";
+import { downTriangle, rightTriangle } from "@library/components/Icons";
+import Button, { ButtonBaseClass } from "@library/components/forms/Button";
+import { t } from "@library/application";
+import TabHandler from "@library/TabHandler";
+
+interface IProps {
+    name: string;
+    className?: string;
+    titleID?: string;
+    children: any[];
+    counter: number;
+    url: string;
+    openParent?: () => void;
+    location: any;
+    depth: number;
+}
+
+interface IState {
+    open: boolean;
+    current: boolean;
+}
+
+export default class SiteNavNode extends React.Component<IProps, IState> {
+    public constructor(props) {
+        super(props);
+        this.state = {
+            open: false,
+            current: false,
+        };
+    }
+
+    public open = (callbackIfAlreadyOpen?: any) => {
+        if (!this.state.open) {
+            this.setState({
+                open: true,
+            });
+        } else {
+            if (callbackIfAlreadyOpen) {
+                callbackIfAlreadyOpen();
+            }
+        }
+    };
+
+    public close = (callbackIfAlreadyClosed?: any) => {
+        if (this.state.open) {
+            this.setState({
+                open: false,
+            });
+        } else {
+            if (callbackIfAlreadyClosed) {
+                callbackIfAlreadyClosed();
+            }
+        }
+    };
+
+    public openRecursive = () => {
+        if (!this.state.current && this.currentPage()) {
+            this.setState({
+                current: true,
+            });
+            if (this.props.openParent) {
+                this.props.openParent();
+            }
+        }
+    };
+
+    public openSelfAndOpenParent = () => {
+        this.setState({
+            open: true,
+        });
+        if (this.props.openParent) {
+            this.open();
+            this.props.openParent();
+        }
+    };
+
+    public toggle = () => {
+        this.setState({
+            open: !this.state.open,
+        });
+    };
+
+    public handleClick = e => {
+        e.preventDefault();
+        this.toggle();
+    };
+
+    public currentPage(): boolean {
+        if (this.props.location && this.props.location.pathname) {
+            return this.props.location.pathname === this.props.url;
```

Comment:
I don't know how much we care about non-canonical URLs, but this won't work with those. Maybe just a note of that.

If it ends up being an issue we'll have to do some parsing of URL and match the record type + record ID on the navigation items here.